src/librbd/AioRequest.cc

```diff
@@ -464,19 +465,22 @@ namespace librbd {
 
   void AbstractWrite::send_write() {
     ldout(m_ictx->cct, 20) << "send_write " << this << " " << m_oid << " "
-			   << m_object_off << "~" << m_object_len << dendl;
-
-    m_state = LIBRBD_AIO_WRITE_FLAT;
-    guard_write();
-    add_write_ops(&m_write);
-    assert(m_write.size() != 0);
+			   << m_object_off << "~" << m_object_len 
+                           << " object exist " << m_object_exist << dendl;
 
-    librados::AioCompletion *rados_completion =
-      librados::Rados::aio_create_completion(this, NULL, rados_req_cb);
-    int r = m_ictx->data_ctx.aio_operate(m_oid, rados_completion, &m_write,
-					 m_snap_seq, m_snaps);
-    assert(r == 0);
-    rados_completion->release();
+    if (m_object_exist) {
+      m_state = LIBRBD_AIO_WRITE_FLAT;
+      guard_write();
+      send_write_op();
+    } else {
+      if (has_parent()) {
+        m_state = LIBRBD_AIO_WRITE_GUARD;
+        complete(-ENOENT);
+      } else {
+        m_state = LIBRBD_AIO_WRITE_FLAT;
```

Comment:
Now that I think about this more, perhaps we should keep the guarded write in the case just to be extra-safe.  That would allow you to combine this block with the "m_object_exists" condition block.